### PR TITLE
Added Kruskal's Algorithm (more organized than the one present)

### PR DESCRIPTION
**Pull request by ruppysuppy**
### **Describe your change:**

* [X] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [X] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [X] This pull request is all my own work -- I have not plagiarized.
* [X] I know that pull requests will not be merged if they fail the automated tests.
* [X] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [X] All new Python files are placed inside an existing directory.
* [X] All filenames are in all lowercase characters with no spaces or dashes.
* [X] All functions and variable names follow Python naming conventions.
* [X] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [X] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [X] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: graphs/minimum_spanning_tree_kruskal2.py [hunk 1]
```diff
@@ -12,17 +15,17 @@ def __init__(self):
         # map from node name to the node object
         self.map = {}
 
-    def make_set(self, x):
+    def make_set(self, x) -> None:
         # create a new set with x as its member
         self.map[x] = Disjoint_Set_Tree_Node(x)
 
-    def find_set(self, x):
+    def find_set(self, x: int) -> Disjoint_Set_Tree_Node:
```
Comment (ruppysuppy) [inline graphs/minimum_spanning_tree_kruskal2.py:22]: Missed the type hints in a few places, will add them

**Comment by ruppysuppy** [comment]
Could you review the PR please, @cclauss?

File: graphs/minimum_spanning_tree_kruskal2.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+    def make_set(self, x: int) -> None:
+        # create a new set with x as its member
+        self.map[x] = Disjoint_Set_Tree_Node(x)
+
+    def find_set(self, x: int) -> Disjoint_Set_Tree_Node:
+        # find the set x belongs to (with path-compression)
+        if self.map[x] != self.map[x].parent:
+            self.map[x].parent = self.find_set(self.map[x].parent.key)
+        return self.map[x].parent
```
Comment (cclauss) [inline graphs/minimum_spanning_tree_kruskal2.py:26]:
```suggestion
        self_map_x = self.map[x] 
        if self_map_x != self_map_x.parent:
            self_map_x.parent = self.find_set(self_map_x.parent.key)
        return self_map_x.parent
```
___OPTIONAL___: You might gain a bit of performance...
https://wiki.python.org/moin/PythonSpeed/PerformanceTips#Loops

Comment (ruppysuppy) [reply graphs/minimum_spanning_tree_kruskal2.py:26]: I'm keeping it as list comprehension, converting
```python
[disjoint_set.make_set(node) for node in self.connections]
````
to
```python
map(disjoint_set.make_set, self.connections.keys())
````
yields:
```python
**********************************************************************
File "d:/Desktop/Projects/GitUploads/Python/graphs/minimum_spanning_tree_kruskal2.py", line 77, in __main__.GraphUndirectedWeighted.kruskal
Failed example:
    mst = graph.kruskal()
Exception raised:
    Traceback (most recent call last):
      File "C:\Program Files\Python37\lib\doctest.py", line 1329, in __run
        compileflags, 1), test.globs)
      File "<doctest __main__.GraphUndirectedWeighted.kruskal[7]>", line 1, in <module>
        mst = graph.kruskal()
      File "d:/Desktop/Projects/GitUploads/Python/graphs/minimum_spanning_tree_kruskal2.py", line 100, in kruskal
        parentu = disjoint_set.find_set(u)
      File "d:/Desktop/Projects/GitUploads/Python/graphs/minimum_spanning_tree_kruskal2.py", line 24, in find_set
        elem_ref = self.map[x]
    KeyError: 1
**********************************************************************
File "d:/Desktop/Projects/GitUploads/Python/graphs/minimum_spanning_tree_kruskal2.py", line 78, in __main__.GraphUndirectedWeighted.kruskal
Failed example:
    assert 5 not in mst.connections[3]
Exception raised:
    Traceback (most recent call last):
      File "C:\Program Files\Python37\lib\doctest.py", line 1329, in __run
        compileflags, 1), test.globs)
      File "<doctest __main__.GraphUndirectedWeighted.kruskal[8]>", line 1, in <module>
        assert 5 not in mst.connections[3]
    NameError: name 'mst' is not defined
**********************************************************************
1 items had failures:
   2 of   9 in __main__.GraphUndirectedWeighted.kruskal
***Test Failed*** 2 failures.
```

File: graphs/minimum_spanning_tree_kruskal2.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        # connections: map from the node to the neighbouring nodes (with weights)
+        # nodes: counting the number of nodes in the graph
+        self.connections = {}
+        self.nodes = 0
+
+    def add_node(self, node: int) -> None:
+        # add a node ONLY if its not present in the graph
+        if node not in self.connections:
+            self.connections[node] = {}
+            self.nodes += 1
```
Comment (cclauss) [inline graphs/minimum_spanning_tree_kruskal2.py:54]: Under what circumstances is self.nodes != len(self.connections)?

Comment (ruppysuppy) [reply graphs/minimum_spanning_tree_kruskal2.py:54]: Yeah its always same. Should I remove it?

Comment (cclauss) [reply graphs/minimum_spanning_tree_kruskal2.py]: Your call.  You already have my approval so you can squash & merge whenever you want (as long as the tests are green).  Well done!

Comment (ruppysuppy) [reply graphs/minimum_spanning_tree_kruskal2.py]: I have made the changes, but I cannot merge since I'm not a member

**Pull request merged**